RUM-7285 Add api-surface step to Lint stage

mariedm opened the pull request:
### What and why?

To ensure public API changes are properly updated, I previously added a reminder in our GitHub PR template checklist to run `make api-surface`. However, this has proven insufficient 🙃 To address this, I am adding an automated step to the CI Lint phase to enforce this check.

### How?

- Add CI Validation:
  - A step is added to the Lint phase that runs `make api-surface`.
  - The generated API surface files are written to temporary files (`api-surface-swift-generated` and `api-surface-objc-generated`).
  - These temporary files are compared with the committed reference files (`api-surface-swift` and `api-surface-objc`).
- CI Behavior:
  - If differences are detected, the CI step exits with an error, prompting contributors to run `make api-surface` locally and commit the updated reference files.
- Updated Reference Files:
  - The reference files are updated as needed to reflect recent changes to the public API.

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes
- [x] Add Objective-C interface for public APIs (see our [guidelines](https://datadoghq.atlassian.net/wiki/spaces/RUMP/pages/3157787243/RFC+-+Modular+Objective-C+Interface#Recommended-solution) [internal]) and run `make api-surface`)

ncreated commented on .gitlab-ci.yml:
```diff
@@ -91,6 +91,17 @@ Lint:
     - make lint license-check
     - make rum-models-verify sr-models-verify
+    - make api-surface ENV=ci
+    - if ! diff api-surface-swift api-surface-swift-generated; then
+        echo "❌ Swift API surface mismatch";
+        echo "👉 Run \`make api-surface\` locally to update \`api-surface-swift\` and commit the changes.";
+        exit 1;
+      fi
+    - if ! diff api-surface-objc api-surface-objc-generated; then
+        echo "❌ Objective-C API surface mismatch";
+        echo "👉 Run \`make api-surface\` locally to update \`api-surface-objc\` and commit the changes.";
+        exit 1;
+      fi
```
suggestion/ Could we move these checks to separate script in `tools/` directory? The convention we started for [.gitlab-ci.yml](https://github.com/DataDog/dd-sdk-ios/blob/develop/.gitlab-ci.yml) is to avoid bare scripts in `yml` as they are very inconvenient to reproduce in local shell.

Ideally, we could get inspired by `make rum-models-verfy`:
* `make api-surface` - to generate API surface files; ran in local
* `make api-surface-verify` - to verify API surface; ran on CI

ncreated commented on Makefile:
```diff
@@ -356,6 +356,16 @@ sr-snapshot-tests-open:
+
 # Generate api-surface files for Datadog and DatadogObjc.
 api-surface:
 		@echo "Generating api-surface-swift"
```
suggestion/ Let's follow the convention in of [printing command's title before it is ran](https://github.com/DataDog/dd-sdk-ios/blob/004ca93c6036c0ceead2f09fedebfae9496c9da3/Makefile#L313). This brings a lot of clarity to CI logs:
```
	@$(ECHO_TITLE) "make api-surface"
```

Also, let's fix the indentation for this command. It uses the old one, all new commands (after migration to GitLab) are indented with less whitespaces. See [these examples](https://github.com/DataDog/dd-sdk-ios/blob/004ca93c6036c0ceead2f09fedebfae9496c9da3/Makefile#L311-L328).

ncreated reviewed (approved):
This is great 🤩! I left two feedbacks on conventions, but well done overall 💪

mariedm commented:
**Update**

- Addressed comments: 
  - Removed inline scripts from `.gitlab-ci.yml` and moved logic to `tools/`
  - Followed command title printing style
  - Corrected `Makefile` indentation
- Split Commands:
  - `Generate` to create API surface files and `Verify` to compare with reference files
  - Added new tests for the new `Verify` command 
- Fixes:
  - Adjusted lint rule in `SessionReplayConfiguration` to fix api-surface file

ncreated commented on tools/api-surface/Tests/APISurfaceCoreTests/IntegrationTests.swift:
```diff
@@ -6,71 +6,155 @@
-
         // Given
-        let command = try SPMLibrarySurfaceCommand.parse([
+        let temporaryFile = "/tmp/api-surface-test-output"
```
suggestion/ Using a dynamically created temp directory could bring some benefits:
- Avoid conflicts between test runs and ensure isolation.
- Automatically clean up after tests to prevent leftover files from affecting others.
- More robust (in other environments) than using hardcoded paths.

Example:
```swift
let tempDirectory = FileManager.default.temporaryDirectory.appendingPathComponent("com.datadoghq.api-surface-" + UUID().uuidString)
try FileManager.default.createDirectory(at: tempDirectory, withIntermediateDirectories: true, attributes: nil)
        
let temporaryFile = tempDirectory.appendingPathComponent("api-surface-test-output").path
```

ncreated reviewed (approved):
Well done 🏅💪!

I did notice however, that integrating `api-surface` has increased the `lint` job duration from 30 seconds to around **3 minutes**. This is a significant change, so before merging, it might be worth discussing with the team:

- How do we feel about this increased duration?
- How might this impact development or the feedback loop?
- Are there any optimization opportunities?
- When should we consider optimizing?

simaoseica-dd reviewed (dismissed):
🏅
This will be very helpful

mariedm commented:
I'm reviving this PR after its priority was bumped; here are the new changes:
- rebased, which includes `v3` release, where each ObjC API has been moved to their feature module (and ObjC module has been removed)
- added a test for ObjC use case (and a fixture)
- updated some dependencies, including `SourceKitten`
- updated tests to use a temporary directory dynamically created for file generation

As discussed, because this step is adding significant time to CI (~3min), we decided to only run it on commits to the `master` branch. This will allow us, when preparing a release, to update the listing of APIs if needed. 

Additionally, I created a [Jira ticket](https://datadoghq.atlassian.net/browse/RUM-11641) listing some potential optimizations. This might allow us to run this step against the `develop` branch as well in future iterations.

maxep reviewed (approved):
LGTM 👍

mariedm commented:
/merge
